### ISSUE TITLE
Revise approach to handling expression type errors

### ISSUE THREAD

**Issue by anandthakker** [mapbox/mapbox-gl-js#5234]
Right now, `coalesce` not only coalesces `null` values, but also serves as a try/catch.  The idea behind this latter behavior was that `coalesce` could be used to provide fallback values for expressions where feature data didn't conform to expected types.  However, @mourner points out that these error-handling semantics are awkward and dangerous (in the sense of swallowing rather than failing hard on errors).  Capturing from chat an revised design for handling mis-typed data:

 - `"get"` returns `null` (rather than producing an error) for a missing property value.
 - Type assertions (`"string"`, `"number"`, ...) and coercions (`"to-string"`, `"to-number"`, ...) accept one or more fallback values in case of mis-typed input, e.g. `["number", ["get", "x"], ["get", "y"], 0]`: if `properties.x` is not numeric, then evaluate to `properties.y` -- if _that's_ not numeric, then evaluate to 0.
 - Either:
   - `"coalesce"` no longer catches errors -- it only coalesces `null`.
   - We remove `"coalesce"` altogether, instead allowing a single, literal fallback value for `"get"` and `"id"`, e.g. `["get", "x", 5]`: evaluate to `properties.x` or, if it is null, to 5.

cc @mourner @jfirebaugh @nickidlugash @aparlato

**Comment by jfirebaugh** [comment]
>  - Either:
>   - `"coalesce"` no longer catches errors -- it only coalesces `null`.
>   - We remove `"coalesce"` altogether, instead allowing a single, literal fallback value for `"get"` and `"id"`, e.g. `["get", "x", 5]`: evaluate to `properties.x` or, if it is null, to 5.

It's probably not a good idea to have two independently-optional arguments to `get`. It's contrived, but consider `["get", "x", ["literal", {"x": "foo"}]`. Is the second argument the object to `get` from, or a fallback for a feature property lookup?

**Comment by anandthakker** [comment]
@jfirebaugh good point.  We could do:

 - `get` no longer defaults to `["properties"]` -- i.e., it requires the object argument: `["get", key, object]` / `["get", key, object, defaultValue]`.
 - introduce `["prop", key]` / `["prop", key, defaultValue]` as an alias for `["get", key, ["properties"]` / `["get", key, ["properties", defaultValue]`

Or we could just keep `coalesce`.  The reason the above alternative is attractive is that it would allow us to remove `Null` altogether.

**Comment by anandthakker** [comment]
Current plan is to go with the first two bullets in the original post, and, for the third bullet, the '"coalesce" no longer catches errors -- it only coalesces null.' option.

**Comment by anandthakker** [comment]
@jfirebaugh thinking about this more, even if we're keeping `Null`, I think it may be worthwhile to introduce `["prop", key]` / `["prop", key, defaultValue]` in lieu of having `get` default to `["properties"]`.  If we don't do this, then users will have to choose between providing default values--via explicit assertion/coercion--and benefitting from type inference:

Expression|Concise|Handles missing value|Handles mistyped value
-----------|--------|----------------------|------------------------
`[curve, [get, 'x'], ...]`| ✅ | ❌| ❌ 
`[curve, [coalesce, [get, 'x'], 0], ...]`| ❌ | ✅ | ❌ 
`[curve, [number, [get, 'x'], 0], ...]`|❌ | ✅| ✅

**Comment by jfirebaugh** [comment]
Eh, I don't think that `[coalesce, [get, 'x'], 0]` or `[number, [get, 'x'], 0]` are too wordy. Think of the slight loss of concision as encouragement to clean up your data so that all features have a value of the same type.

**Comment by anandthakker** [comment]
> Think of the slight loss of concision as encouragement to clean up your data so that all features have a value of the same type.

Heh 😄.  I _almost_ agree with this -- the sticking point for me is a property value that is conceptually typed `optional<Number>`.  It's not unreasonable to use `null` to mean `None` in that case.   But I'll concede that `[number, [get, 'x'], 0]` isn't all that wordy.

**Comment by nickidlugash** [comment]
> the sticking point for me is a property value that is conceptually typed optional<Number>. It's not unreasonable to use null to mean None in that case. But I'll concede that [number, [get, 'x'], 0] isn't all that wordy.

We commonly use `null` to mean None in our Mapbox Streets vector tiles, for both numbers and string fields. I don't have strong objections against `[coalesce, [get, 'x'], 0]` or `[number, [get, 'x'], 0]`, but just wanted to flag that this will be a fairly common syntax pattern in our styles.
